### PR TITLE
Support for multilang ZIMs

### LINKED ISSUE #903: ZIM with multiple language metadata are not properly supported
At least not supported in kiwix-serve filter

Screenshot 2023-02-22 at 09 50 22

First reported at https://github.com/kiwix/libkiwix/issues/872#issuecomment-1439722930

### PR DESCRIPTION
**Pull request by veloman-yunkan**
Fixes #903

Note that a multilanguage ZIM/book is counted as 1 full book in the results to each of the matching `/catalog/v2/entries?lang=<LANG>` queries.

### CODE REVIEW

**Comment by mgautierfr** [comment]
While this PR indeed fix #903, it seems a bit too simple to fully support multilang ZIMs. (And I would be pleasantly surprised if it is enough)

At least, in `Library::updateBookDB`, we use `book.getLanguage()` to set the stemmer. With a multilang book, it will not work.
The same way, we don't split the comma separated list when indexing the book itself. It seems to work (based on the tests you have added), propably because we find the queried language in the book's language list but it would be better to explicitly split it.

And I think we should change `std::string Book::getLanguage()` to `std::vector<std::string> Book::getLanguages()`. This way, we would force all our users to handle correctly multi language book.

**Comment by veloman-yunkan** [comment]
> in `Library::updateBookDB`, we use `book.getLanguage()` to set the stemmer. With a multilang book, it will not work.

What stemmer should be used for a multilang book? Currently, in such cases the stemmer is simply not used since an attempt to create a stemmer for a non-existing language fails with an exception which is intercepted and ignored.

> The same way, we don't split the comma separated list when indexing the book itself. It seems to work (based on the tests you have added), propably because we find the queried language in the book's language list but it would be better to explicitly split it.

Will fix.

> And I think we should change `std::string Book::getLanguage()` to `std::vector<std::string> Book::getLanguages()`

Should we do that in one step? Or add `Book::getLanguages()` and deprecate `Book::getLanguage()`?

**Comment by veloman-yunkan** [comment]
> > And I think we should change `std::string Book::getLanguage()` to `std::vector<std::string> Book::getLanguages()`
> 
> Should we do that in one step? Or add `Book::getLanguages()` and deprecate `Book::getLanguage()`?

Another option is to preserve `Book::getLanguage()` as a wrapper around `Book::getLanguages()` that throws when called on a multilang book.

**Comment by mgautierfr** [comment]
Agree with the current code.

I think it would be better to depreciate `Book::getLanguage`. We can be sure that people will forget to catch the potential exception thrown if we are in multizim. It is better to move the change explicitly in the api.

**Comment by veloman-yunkan** [comment]
> I think it would be better to depreciate `Book::getLanguage`. We can be sure that people will forget to catch the potential exception thrown if we are in multizim. It is better to move the change explicitly in the api.

@mgautierfr OK, I will deprecate it. Should its current implementation remain unchanged (i.e. should it keep returning a comma separated list of languages for multilang ZIMs)? The other alternatives are:

1. return the first element of `Book::getLanguages()` (or an empty string if the latter is empty)
2. return the only element of `Book::getLanguages()` if there is only one language, otherwise raise an exception.

**Comment by mgautierfr** [comment]
> Should its current implementation remain unchanged (i.e. should it keep returning a comma separated list of languages for multilang ZIMs) ?

Yes. We keep the function (as deprecated) to not break the api, so we shouldn't change the behavior (even if it is buggy)

> And what about Bookmark::getLanguage()? Should we make the same change to Bookmark?

Bookmark is about only one article. We can expect that even if zim file is multilanguages, one article is in one language only, so no change needed.

File: src/library.cpp [hunk 1]
```diff
@@ -455,12 +455,14 @@ void Library::updateBookDB(const Book& book)
...
-    indexer.set_stemming_strategy(Xapian::TermGenerator::STEM_SOME);
-  } catch (...) {}
+  const auto langs = book.getLanguages();
+  if ( langs.size() == 1 ) {
+    try {
+      stemmer = Xapian::Stem(iso639_3ToXapian(langs[0]));
+      indexer.set_stemmer(stemmer);
+      indexer.set_stemming_strategy(Xapian::TermGenerator::STEM_SOME);
+    } catch (...) {}
+  }
```
Comment (mgautierfr) [inline src/library.cpp:465]: Could we try to load languages until we found one (not only checking the first one) ?

File: src/server/internalServer.cpp [hunk 2]
```diff
@@ -223,7 +223,8 @@ typedef std::set<std::string> Languages;
 Languages getLanguages(const Library& lib, const Library::BookIdSet& bookIds) {
   Languages langs;
   for ( const auto& b : bookIds ) {
-    langs.insert(lib.getBookById(b).getLanguage());
+    const auto bookLangs = lib.getBookById(b).getLanguages();
+    langs.insert(bookLangs.begin(), bookLangs.end());
```
Comment (mgautierfr) [inline src/server/internalServer.cpp:227]: This seems a functional change, not just a deprecation of `getLanguage` so it should go in another commit.
(Not important in itself, but as we have other changes to do, let's split this change correctly)

**Review by mgautierfr — CHANGES_REQUESTED** [review]

File: src/library.cpp (hunk 1 above)
Comment (veloman-yunkan) [reply src/library.cpp:465]: I think you missed the `if()` surrounding the `try{}catch()`. A language (and a corresponding stemmer) is associated with the book metadata (title, description, tags, etc) only if the book is known to be in one language. Otherwise, if the book is a multilanguage one how can we figure out in which language, say, its description is written?

File: src/server/internalServer.cpp (hunk 2 above)
Comment (veloman-yunkan) [reply src/server/internalServer.cpp:227]: Done

File: src/library.cpp (hunk 1 above)
Comment (mgautierfr) [reply src/library.cpp:465]: I haven't missed the `if()`. But I have missed the fact it was at indexing time. I was missing with libzim feature of detect the query language based on the database language.

I was totally wrong on this side. So we are good here 🙂

**Review by mgautierfr — APPROVED** [review]
